Cython/Compiler/ExprNodes.py

```diff
@@ -2908,67 +2903,66 @@ def analyse_types(self, env):
-            if self.module_names:
-                assert self.module_names.is_literal  # make sure we create the tuple only once
-            if self.level == 0:
-                utility_code = UtilityCode.load_cached("ImportDottedModule", "ImportExport.c")
-                helper_func = "__Pyx_ImportDottedModule"
-            else:
-                utility_code = UtilityCode.load_cached("ImportDottedModuleRelFirst", "ImportExport.c")
-                helper_func = "__Pyx_ImportDottedModuleRelFirst"
-            code.globalstate.use_utility_code(utility_code)
-            import_code = "%s(%s, %s, %s)" % (
-                helper_func,
-                self.module_name.py_result(),
-                self.module_names.py_result() if self.module_names else 'NULL',
-                self.module_qualname.py_result(),
-            )
-        else:
-            code.globalstate.use_utility_code(UtilityCode.load_cached("Import", "ImportExport.c"))
-            import_code = "__Pyx_Import(%s, %s, %d, %s)" % (
-                self.module_name.py_result(),
-                self.name_list.py_result() if self.name_list else '0',
-                self.level,
-                self.module_qualname.py_result())
 
+        module_qualname = 'NULL'
+        if self.module_qualname is not None:
+            module_qualname = code.get_py_string_const(self.module_qualname)
+
+        code.globalstate.use_utility_code(UtilityCode.load_cached("Import", "ImportExport.c"))
+        import_code = "__Pyx_Import(%s, %s, %s, %d)" % (
+            self.module_name.py_result(),
+            self.name_list.py_result() if self.name_list else '0',
+            module_qualname,
+            self.level)
+        tmp_result = code.funcstate.allocate_temp(self.type, manage_ref=False)
         code.putln("%s = %s; %s" % (
-            self.result(),
+            tmp_result,
             import_code,
-            code.error_goto_if_null(self.result(), self.pos)))
+            code.error_goto_if_null(tmp_result, self.pos)))
+
+        if self.as_name and "." in self.module_name.value:
+            # We need to get the submodules in this case
+            code.globalstate.use_utility_code(UtilityCode.load_cached("ImportFrom", "ImportExport.c"))
+            submodule = code.funcstate.allocate_temp(self.type, manage_ref=False)
+            modules = self.module_name.value.split(".")
+            for module in modules[1:]:
+                module_obj = code.get_py_string_const(StringEncoding.EncodedString(module))
+                code.putln("%s = __Pyx_ImportFrom(%s, %s);" % (
+                    submodule,
+                    tmp_result,
+                    module_obj))
+                code.putln("if (unlikely(!%s)) { Py_DECREF(%s); %s; }" % (
+                    submodule,
+                    tmp_result,
+                    code.error_goto(self.pos)))
+                code.putln("Py_DECREF(%s); %s = %s;" % (tmp_result, tmp_result, submodule))
```

Comment:
```suggestion
                code.putln(f"Py_DECREF(tmp_result);")
                code.putln(code.error_goto_if_null(submodule, self.pos))
                code.putln(f"{tmp_result} = {submodule};")
```

(The same thing but slightly simpler I think)